src/SAWScript/Crucible/MIR/TypeShape.hs

```diff
@@ -0,0 +1,274 @@
+        M.TyTuple tys -> goTuple ty tys
+        M.TyClosure tys -> goTuple ty tys
+        M.TyFnDef _ -> goUnit ty
+        M.TyArray ty' _ | Some shp <- go ty' -> Some $ ArrayShape ty ty' shp
+        M.TyAdt nm _ _ -> case Map.lookup nm (col ^. M.adts) of
+            Just adt | Just ty' <- reprTransparentFieldTy col adt ->
+                mapSome (TransparentShape ty) $ go ty'
+            Just (M.Adt _ M.Struct [v] _ _ _ _) -> goStruct ty (v ^.. M.vfields . each . M.fty)
+            Just (M.Adt _ ak _ _ _ _ _) -> error $ "tyToShape: AdtKind " ++ show ak ++ " NYI"
+            Nothing -> error $ "tyToShape: bad adt: " ++ show ty
+        M.TyRef ty' mutbl -> goRef ty ty' mutbl
+        M.TyRawPtr ty' mutbl -> goRef ty ty' mutbl
+        M.TyFnPtr sig -> goFnPtr ty sig
+        _ -> error $ "tyToShape: " ++ show ty ++ " NYI"
+
+    goPrim :: M.Ty -> Some TypeShape
+    goPrim ty | Some tpr <- tyToRepr col ty, AsBaseType btpr <- asBaseType tpr =
+        Some $ PrimShape ty btpr
+    goPrim ty | Some tpr <- tyToRepr col ty =
+        error $ "tyToShape: type " ++ show ty ++ " produced non-primitive type " ++ show tpr
+
+    goUnit :: M.Ty -> Some TypeShape
+    goUnit ty = Some $ UnitShape ty
+
+    goTuple :: M.Ty -> [M.Ty] -> Some TypeShape
+    goTuple ty tys | Some flds <- loop tys Empty = Some $ TupleShape ty tys flds
+      where
+        loop :: forall ctx. [M.Ty] -> Assignment FieldShape ctx -> Some (Assignment FieldShape)
+        loop [] flds = Some flds
+        loop (ty':tys') flds | Some fld <- go ty' = loop tys' (flds :> OptField fld)
+
+    goStruct :: M.Ty -> [M.Ty] -> Some TypeShape
+    goStruct ty tys | Some flds <- loop tys Empty = Some $ StructShape ty tys flds
+      where
+        loop :: forall ctx. [M.Ty] -> Assignment FieldShape ctx -> Some (Assignment FieldShape)
+        loop [] flds = Some flds
+        loop (ty':tys') flds | Some fld <- goField ty' = loop tys' (flds :> fld)
+
+    goField :: M.Ty -> Some FieldShape
+    goField ty | Some shp <- go ty = case canInitialize col ty of
+        True -> Some $ ReqField shp
+        False -> Some $ OptField shp
+
+    goRef :: M.Ty -> M.Ty -> M.Mutability -> Some TypeShape
+    goRef ty ty' mutbl
+      | M.TySlice slicedTy <- ty'
+      , Some tpr <- tyToRepr col slicedTy
+      = Some $
+         TupleShape ty [refTy, usizeTy]
+             (Empty
+                :> ReqField (RefShape refTy slicedTy mutbl tpr)
+                :> ReqField (PrimShape usizeTy BaseUsizeRepr))
+      | M.TyStr <- ty'
+      = Some $
+        TupleShape ty [refTy, usizeTy]
+            (Empty
+                :> ReqField (RefShape refTy (M.TyUint M.B8) mutbl (BVRepr (knownNat @8)))
+                :> ReqField (PrimShape usizeTy BaseUsizeRepr))
+      where
+        -- We use a ref (of the same mutability as `ty`) when possible, to
+        -- avoid unnecessary clobbering.
+        refTy = case ty of
+            M.TyRef _ _ -> M.TyRef ty' mutbl
+            _ -> M.TyRef ty' mutbl
```

Comment:
The result of these cases is the same. Is that intentional?

Reply:
While I didn't write this code (it was moved from a different module in `crucible-mir-comp`), I agree that this code is rather silly as written. I've pushed a separate commit to remove the unnecessary `case` expression.